### PR TITLE
TA-453- Adding crossbrowser files and dependencies

### PR DESCRIPTION
**Pull request by Rob-Doherty**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/TA-453

### Change description ###

Adding needed files and config to run cross-browser tests in the nightly pipeline on all supported desktop browsers.

Includes:
- run the @smoke-tests tagged e2e test as the crossbrowser test
- update Jenkinsfile_nightly to run the crossbrowser test on all supported desktop browsers
- refactor the e2e test helpers to support both Puppeteer and WebDriver
- new supportedBrowsers.js, a list of supported browsers
- new SauceLabsReportingHelper.js file to update Saucelabs with the result of the test, and to integrate with the Saucelabs reports and videos in Jenkins
- new saucelabs.conf.js file to configure the Codecept test run with parameters relevant to the corssbrwoser tests, e.g. slower timeouts to allow for network delays
- updated package.json file with new `test:crossbrowser` call and WebdriverIO dependency

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by Rob-Doherty** [comment]
Evidence of crossbrowser changes going green in nightly build:
image

File: e2e/crossbrowser/supportedBrowsers.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      'sauce:options': {
+        name: 'FPLA: IE11',
+        screenResolution: '1400x1050',
+      },
+    },
+    edge_win_latest: {
+      browserName: 'MicrosoftEdge',
+      platformName: LATEST_WINDOWS,
+      browserVersion: 'latest',
+      'sauce:options': {
```
Comment (imran-hmcts) [inline e2e/crossbrowser/supportedBrowsers.js:19]: More of a question than a comment. IIRC, the resolution defaults to something minimal like 1024x768 when it isn't specified. Is this acceptable or expected test parameter?

File: e2e/helpers/dump_browser_logs_helper.js [hunk 2]
```diff
@@ -28,7 +28,8 @@ function stringify(value) {
 
 module.exports = class HooksHelpers extends Helper {
   async _failed(test) {
-    const logs = (await this.helpers['Puppeteer'].grabBrowserLogs()).map(log => {
+    const helper = this.helpers['Puppeteer'] || this.helpers['WebDriver'];
```
Comment (imran-hmcts) [inline e2e/helpers/dump_browser_logs_helper.js:31]: Minor thing. This line repeats a lot in the file so maybe refactor it in a common function?

File: e2e/crossbrowser/supportedBrowsers.js (hunk 1 above)
Comment (Rob-Doherty) [reply e2e/crossbrowser/supportedBrowsers.js:19]: good question @imran-hmcts, so the way this config works is that it overwrites a default config defined in `saucelabs.conf.js`
```
const defaultSauceOptions = {
  username: process.env.SAUCE_USERNAME,
  accessKey: process.env.SAUCE_ACCESS_KEY,
  tunnelIdentifier: process.env.TUNNEL_IDENTIFIER || 'reformtunnel',
  acceptSslCerts: true,
  windowSize: '1600x900',
  tags: ['FPL']
};
```
So by default the window size is `1600x900` but we can override that here

File: e2e/helpers/dump_browser_logs_helper.js (hunk 2 above)
Comment (Rob-Doherty) [reply e2e/helpers/dump_browser_logs_helper.js:31]: great idea, done.

Comment (Rob-Doherty) [reply e2e/helpers/dump_browser_logs_helper.js:31]: I've added a function in browser_helper.js where there were multiple repeating lines

File: e2e/helpers/SauceLabsReportingHelper.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+const event = require('codeceptjs').event;
+const container = require('codeceptjs').container;
+const exec = require('child_process').exec;
+
+function updateSauceLabsResult(result, sessionId) {
+  console.log('SauceOnDemandSessionID=' + sessionId + ' job-name=probate-frontend');
```
Comment (Rob-Doherty) [inline e2e/helpers/SauceLabsReportingHelper.js:8]: update!!

**Review by lewisbirks — DISMISSED** [review]

**Review by lewisbirks — APPROVED** [review]

**Pull request merged**